Objects/frameobject.c

```diff
@@ -1707,70 +1778,94 @@ frame_setlineno(PyObject *op, PyObject* p_new_lineno, void *Py_UNUSED(closure))
     while (start_stack > best_stack) {
         if (top_of_stack(start_stack) == Except) {
             /* Pop exception stack as well as the evaluation stack */
-            PyObject *exc = PyStackRef_AsPyObjectBorrow(_PyFrame_StackPop(f->f_frame));
+            PyObject *exc = PyStackRef_AsPyObjectBorrow(_PyFrame_StackPop(self->f_frame));
             assert(PyExceptionInstance_Check(exc) || exc == Py_None);
             PyThreadState *tstate = _PyThreadState_GET();
             Py_XSETREF(tstate->exc_info->exc_value, exc == Py_None ? NULL : exc);
         }
         else {
-            PyStackRef_XCLOSE(_PyFrame_StackPop(f->f_frame));
+            PyStackRef_XCLOSE(_PyFrame_StackPop(self->f_frame));
         }
         start_stack = pop_value(start_stack);
     }
     /* Finally set the new lasti and return OK. */
-    f->f_lineno = 0;
-    f->f_frame->instr_ptr = _PyFrame_GetBytecode(f->f_frame) + best_addr;
+    self->f_lineno = 0;
+    self->f_frame->instr_ptr = _PyFrame_GetBytecode(self->f_frame) + best_addr;
     return 0;
 }
 
+/*[clinic input]
+@critical_section
+@getter
+frame.f_trace as frame_trace
+
+Return the trace function for this frame, or None if no trace function is set.
+[clinic start generated code]*/
+
 static PyObject *
-frame_gettrace(PyObject *op, void *Py_UNUSED(closure))
+frame_trace_get_impl(PyFrameObject *self)
+/*[clinic end generated code: output=5475cbfce07826cd input=f382612525829773]*/
 {
-    PyFrameObject *f = PyFrameObject_CAST(op);
-    PyObject* trace = f->f_trace;
-    if (trace == NULL)
+    PyObject* trace = self->f_trace;
+    if (trace == NULL) {
         trace = Py_None;
+    }
     return Py_NewRef(trace);
 }
 
+/*[clinic input]
+@critical_section
+@setter
+frame.f_trace as frame_trace
+[clinic start generated code]*/
+
 static int
-frame_settrace(PyObject *op, PyObject* v, void *Py_UNUSED(closure))
+frame_trace_set_impl(PyFrameObject *self, PyObject *value)
+/*[clinic end generated code: output=d6fe08335cf76ae4 input=d96a18bda085707f]*/
 {
-    PyFrameObject *f = PyFrameObject_CAST(op);
-    if (v == Py_None) {
-        v = NULL;
+    if (value == Py_None) {
+        value = NULL;
     }
-    if (v != f->f_trace) {
-        Py_XSETREF(f->f_trace, Py_XNewRef(v));
-        if (v != NULL && f->f_trace_opcodes) {
-            return _PyEval_SetOpcodeTrace(f, true);
+    if (value != self->f_trace) {
+        Py_XSETREF(self->f_trace, Py_XNewRef(value));
+        if (value != NULL && self->f_trace_opcodes) {
+            return _PyEval_SetOpcodeTrace(self, true);
         }
     }
     return 0;
 }
 
+/*[clinic input]
+@critical_section
+@getter
+frame.f_generator as frame_generator
```

Comment:
Since this uses critical section, the corresponding C API `PyFrame_GetGenerator` should also use it.